modules/multiprocessExample/main.py

```diff
@@ -0,0 +1,160 @@
+        # HOWEVER: Sharing memory between processes is a pain,
+        # and the worst case is that we end up
+        # processing an extra item before exiting and dying
+        if (exitRequested(exitRequest)):
+            break
+
+        # Better example of the above (C++ pseudo-code):
+
+        # bool exitRequest; // Global
+        # mutexlock exitMtx; // Global
+        # ...
+        # exitMtx.lock(); // Entering critical section
+        # if (exitRequest) {
+        #     exitMtx.unlock(); // Leaving critical section
+        #     break; // Exiting loop, so we won't accidentally unlock twice
+        # }
+        # exitMtx.unlock() // Leaving critical section
+
+    # Once the process reaches the end of the function it will die automatically
+    print("Counter finished!")
+
+
+if __name__ == "__main__":
+
+    # This is what I will call the manager
+    # The manager is responsible for any supporting locks and queues
+    # and creating and initializing class objects
+    # IMPORTANT: Class globals are NOT retained across processes and setting them at runtime is HIGHLY DISCOURAGED
+    counters = [
+        counter.Counter(1),
+        counter.Counter(4),
+        counter.Counter(3)
+    ]
+    counterPause = mp.Lock()
+    counterToMainPipeline = mp.Queue()
+    counterStop = mp.Queue(1)  # Queue of size 1
+
+    numProcesses = 3
+
+    # Prepare processes
+    p = []
+    for i in range(0, numProcesses):
+        p.append(mp.Process(target=counterWorker,
+                            args=(counters[i], counterPause, counterStop, counterToMainPipeline,)))
+
+    # Start processes
+    for i in range(0, numProcesses):
+        p[i].start()
+
+    # Run for some time and then stop
+    time.sleep(2)
+    counterPause.acquire()
+    print("Paused")
+    time.sleep(4)
+    print("Resumed")
+    counterPause.release()
+    time.sleep(2)
+
+    # Stop must be requested while the processes are running!
+    # "Well, why not release the pause lock then?"
+    # What if it isn't paused? Double release will cause an exception
+    # "Okay, well if it isn't paused that's okay, just catch the exception"
+    # This adds complexity to both the manager and the process function, which is bad
+    # "Not as bad as crash, also it will catch double resumes"
+    # Silent error is pretty bad, especially if the manager acquires a pause lock while
+    # a process is still inside the (empty) critical section. Can you see how this would happen?
+    #  - Xierumeng's brain
+    counterStop.put(True)  # Any item will do, but True is used here to make it clearer
+
+    # In case the processes are stuck on a queue
+    # The maximum number of processes stuck on a single in or out queue
+    # is the number of running processes (obviously)
+    # They may all be waiting for data (stuck on empty queue), so we put in that number of elements
+    # They may all be trying to output data (stuck on full queue), so we pop that number of elements
```

Comment:
Add that losing a bit of data is not the end of the world, especially since we're halting the entire thing anyway.